tests/framework/artifacts.py

```diff
@@ -0,0 +1,191 @@
+    ARTIFACTS_ROOT = 'ci-artifacts'
+    ARTIFACTS_DISKS = '/disks/' + platform.machine()
+    ARTIFACTS_KERNELS = '/kernels/' + platform.machine()
+    ARTIFACTS_MICROVMS = '/microvms'
+    ARTIFACTS_SNAPSHOTS = '/snapshots/' + platform.machine()
+
+    def __init__(
+        self,
+        bucket
+    ):
+        """Initialize S3 client."""
+        config = botocore.client.Config(signature_version=botocore.UNSIGNED)
+        # pylint: disable=E1101
+        # fixes "E1101: Instance of '' has no 'Bucket' member (no-member)""
+        self.bucket = boto3.resource('s3', config=config).Bucket(bucket)
+
+    def _fetch_artifacts(self,
+                         artifact_root,
+                         artifact_dir,
+                         artifact_ext,
+                         artifact_type,
+                         artifact_class,
+                         keyword=None):
+        artifacts = []
+        prefix = artifact_root + artifact_dir
+        files = self.bucket.objects.filter(Prefix=prefix)
+        for file in files:
+            if (
+                # Filter by extensions.
+                file.key.endswith(artifact_ext)
+                # Filter by userprovided keyword if any.
+                and (keyword is None or keyword in file.key)
+            ):
+                artifacts.append(artifact_class(self.bucket,
+                                                file.key,
+                                                artifact_type=artifact_type))
+        return artifacts
+
+    def microvms(self, keyword=None):
+        """Return microvms artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_MICROVMS,
+            ArtifactCollection.MICROVM_CONFIG_EXTENSION,
+            ArtifactType.MICROVM,
+            Artifact,
+            keyword=keyword
+        )
+
+    def kernels(self, keyword=None):
+        """Return kernel artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_KERNELS,
+            ArtifactCollection.MICROVM_KERNEL_EXTENSION,
+            ArtifactType.KERNEL,
+            Artifact,
+            keyword=keyword
+        )
+
+    def disks(self, keyword=None):
+        """Return disk artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_DISKS,
+            ArtifactCollection.MICROVM_DISK_EXTENSION,
+            ArtifactType.DISK,
+            DiskArtifact,
+            keyword=keyword
+        )
+
+
+class ArtifactSet:
+    """Manages a set of artifacts with the same type."""
+
+    def __init__(self, artifacts):
+        """Initialize type and artifact array."""
+        assert len(artifacts) > 0
+        self._type = artifacts[0].type
+        self._artifacts = []
+        self.add(artifacts)
```

Comment:
IMO it seems a little bit weird to have an `add` function called in an initialization function. If `add()` is only used here can we give up on this function and insert here, those artifacts (this way the second `assert` is not required)?
If we need in the future a function that adds new artifacts to the set we can make a separate one that will be used for this scope, i.e. adding (appending) not initializing :-??

Reply:
`add()` contains logic that guarantees all artifacts in set are the same. It can be called to add more artifacts to the set after construction. Why would I duplicate it ?

Reply:
Renamed to insert().